src/pip/_internal/resolution/resolvelib/found_candidates.py

```diff
@@ -32,18 +31,21 @@ def _insert_installed(installed, others):
     already-installed package. Candidates from index are returned in their
     normal ordering, except replaced when the version is already installed.
 
-    Since candidates from index are already sorted by reverse version order,
-    `sorted()` here would keep the ordering mostly intact, only shuffling the
-    already-installed candidate into the correct position. We put the already-
-    installed candidate in front of those from the index, so it's put in front
-    after sorting due to Python sorting's stableness guarentee.
+    The implementation iterates through and yields other candidates, inserting
+    the installed candidate exactly once before we start yielding older or
+    equivalent candidates, or after all other candidates if they are all newer.
     """
-    candidates = sorted(
-        itertools.chain([installed], others),
-        key=operator.attrgetter("version"),
-        reverse=True,
-    )
-    return iter(candidates)
+    installed_yielded = False
+    for candidate in others:
+        # If the installed candidate is better, yield it first.
+        if not installed_yielded and installed.version >= candidate.version:
+            yield installed
+            installed_yielded = True
+        yield candidate
+
```

Comment:
So the key problem with #9264 was that the `sorted(...)` approach didn't work, correct? I assume that was because the assumption "candidates from index are already sorted by reverse version order" was somehow incorrect - is that right?

The logic here seems fine, but so does the original logic - so I'm a bit reluctant to approve without understanding how this fixes the problem that caused #9264 to be reverted. Can you explain a bit more? I assume it was the "upgrading pip goes into an infinite loop" thing, but I don't see how this change improves that situation.

Sorry if I'm being dense here...

Reply:
The sorted approach works (in a sense), and the assumption it made is correct. The problem is it consumes the lazy iterator eagerly, and when it does that the preparer downloads all the distributions from PyPI. This implementation tries to implement the same logic with a generator to avoid that eager consumption.

Reply:
Notably, that eager consumption meant that we'd download all versions of pip before returning from this function. :)

Reply:
Ah! Right, I get it now.

Maybe add a comment here saying that it's important to only consume items from `others` when needed, as consuming an item prepares the distribution? It seems to me that this would be an easy mistake to make, and the more places in the code we can add reminders, the better.

Reply:
I added a module-level docstring to explain how it’s important to keep things in `FoundCandidates` lazy to not kill performance.